### PR TITLE
Manage catcher trails by lifetime entries 

### PR DESCRIPTION
**Pull request by ekrctb**
Catcher trails were missing when a replay is rewound.
With this PR, catcher trails are displayed after a replay is rewound.

A concern is the overhead of the allocation of entries. The number of entries can go as high as 10k for a long map with a lot of hyper fruits. However, I found no visible performance differences.
Also, I think if the storing entries ever became an issue, the lifetime entry container can implement a way to forget old entries when not needed (if rewinding is not allowed).

### CODE REVIEW

File: osu.Game.Rulesets.Catch/UI/CatcherArea.cs [hunk 1]
```diff
@@ -109,15 +109,23 @@ protected override void UpdateAfterChildren()
 
             comboDisplay.X = Catcher.X;
 
-            if (!lastHyperDashState && Catcher.HyperDashing && Time.Elapsed > 0)
-                catcherTrails.DisplayHyperDashAfterImage(Catcher.CurrentState, Catcher.X, Catcher.BodyScale);
+            if (Time.Elapsed <= 0)
+            {
+                // This is probably a wrong value, but currently the true value is not recorded.
+                // Setting `true` will prevent generation of false-positive after-images (with more false-positives).
+                lastHyperDashState = true;
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/CatcherArea.cs:116]: Can you explain how this false-positive generation happens? I'd like to think the negative rate/elapsed check should suffice to ensure that no extra entries are added during a rewind...?

File: osu.Game.Rulesets.Catch/UI/CatcherTrail.cs [hunk 2]
```diff
@@ -34,11 +29,40 @@ public CatcherTrail()
             };
         }
 
-        protected override void FreeAfterUse()
+        protected override void OnApply(CatcherTrailEntry entry)
         {
-            ClearTransforms();
+            Scale = entry.Scale;
+            Position = new Vector2(entry.Position, 0);
             Alpha = 1;
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/CatcherTrail.cs:36]: Other places do this by doing `ApplyTransformsAt(double.MinValue)` in `OnFree()` before clearing transforms, but I suppose this also works.

**Review by bdach — COMMENTED** [review]
Looks mostly okay.

File: osu.Game.Rulesets.Catch/UI/CatcherArea.cs (hunk 1 above)
Comment (ekrctb) [reply osu.Game.Rulesets.Catch/UI/CatcherArea.cs:116]: The comment is wrong and

```suggestion
                // Setting `true` will prevent generation of false-positive after-images (with more false-negatives).
                lastHyperDashState = true;
```

And, without `lastHyperDashState = true`, false positives happen.
When a hyper-fruit is caught when the catcher was already hyper-dashing, no afterimage should be displayed. But `lastHyperDashState` could be false when a replay is rewound to the time between the first hyper-fruit and the second hyper-fruit.

Comment (bdach) [reply osu.Game.Rulesets.Catch/UI/CatcherArea.cs]: Ok, I see, that explains it, thanks.

Mostly spitballing here, but could this instead be fixed by maintaining a "last trail spawned" time, and never spawning trails if the current time is less than that last trail spawn time? The rationale in my head is that if an afterimage entry has been spawned for a hitobject already prior to a rewind, there is no point in doing any more forward processing again *after* the rewind (at least until the gameplay catches up to where it was previously).

Comment (ekrctb) [reply osu.Game.Rulesets.Catch/UI/CatcherArea.cs]: That works for a replay, but all other codes don't assume the play is not changing after a rewind.
Specifically, the replay can change after a rewind in the current editor (consider: rewind -> edit -> play).
Also, there is a future possibility of something like a quicker quick-retry, I imagine.
Note that it is specifying `RemoveRewoundEntry` to remove entries when rewound.

Comment (bdach) [reply osu.Game.Rulesets.Catch/UI/CatcherArea.cs]: Ok, I guess that doesn't universally work. Probably fine to let be for now, and eventually the info about the hyperdash trigger time should be stored to the judgements, probably. mania does similar things with similar effects from what I recall.

**Review by bdach — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]
lgtm

**Pull request merged**